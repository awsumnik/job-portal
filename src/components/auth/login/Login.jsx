import React, { useEffect, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import Button from '../../common/button/Button';
import Card from '../../common/card/Card';
import Input from '../../common/input/Input';
import { StyledCenteredWrapper } from '../../common/styled/StyledComponents';
import { DEFAULT_ROUTES } from '../../../constants';
import { handleLoginAction } from '../../../reducers/user/actionCreator';
import messages from '../../../nls/en.json';

const Login = () => {
  const dispatch = useDispatch();
  const navigate = useNavigate();

  const { user: { loading, error, userType, isLoggedIn } } = useSelector(state => state);
  const initialFormData = Object.freeze({
    email: "",
    password: "",
  });
  const [formErrors, setFormErrors] = useState({
    email: true,
    password: true,
  });
  const [formData, updateFormData] = useState(initialFormData);

  const handleChange = (e) => {
    updateFormData({
      ...formData,
      [e.target.name]: e.target.value.trim()
    });
  };

  const onFormError = (errObj) => {
    setFormErrors({
      ...formErrors,
      ...errObj,
    });
  };

  const hasError = () => Object.values(formErrors).some(el => el);

  useEffect(() => {
    if (isLoggedIn) {
      const redirectPath = DEFAULT_ROUTES[userType];
      navigate(redirectPath);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isLoggedIn]);

  const handleLoginClick = (event) => {
    event.preventDefault();
    dispatch(handleLoginAction(formData));
  };

  const goToSignUp = () => {
    navigate('/sign-up');
  };

  return (
    <StyledCenteredWrapper>
      <form className='form' autoComplete='off' onSubmit={handleLoginClick}>
        <Card style={{ width: 400 }}>
          <div className="card-title text-center">{ messages.LOGIN_HEADING }</div>
          {
            error && !loading && (
              <div className='error-text text-center'>{error}</div>
            )
          }
          <div className="card-content">
            <div>
              <Input
                type="email"
                label="Email"
                name="email"
                placeholder={'Enter Email'}
                onChange={handleChange}
                isRequired
                minLength={8}
                autoFocus
                isEmail
                onError={onFormError}
              />
            </div>
            <div>
              <Input
                type="password"
                label="Password"
                name="password"
                placeholder={'Enter Password'}
                onChange={handleChange}
                isRequired
                isPassword
                minLength={8}
                onError={onFormError}
              />
            </div>
            <div>
              <Button
                type="submit"
                label="Log in"
                name="loginBtn"
                priority="primary"
                onClick={handleLoginClick}
                isDisabled={hasError()}
              />
            </div>
            <div>
              <hr />
              <p className='text-center'>{ messages.DONT_HAVE_ACCOUNT }</p>
              <Button
                type="button"
                label="Sign Up"
                name="signUpBtn"
                priority="secondary"
                onClick={goToSignUp}
              />
            </div>
          </div>
        </Card>
      </form>
    </StyledCenteredWrapper>
  )
}

export default Login;