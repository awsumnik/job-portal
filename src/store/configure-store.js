import { applyMiddleware, createStore } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import thunk from 'redux-thunk';
import rootReducer from '../reducers';
import { loadInititalAppState, savedState, saveState } from './';
loadInititalAppState()

let store;
if (process.env.NODE_ENV === 'development') {
  store = createStore(
    rootReducer,
    savedState,
    composeWithDevTools(applyMiddleware(thunk)),
  );
} else {
  store = createStore(rootReducer, savedState, applyMiddleware(thunk));
}

store.subscribe(() => {
  saveState(store.getState());
});

export default store;