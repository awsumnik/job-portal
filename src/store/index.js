import { themeIntitialState } from '../reducers/theme/themeReducer';
import { userInitialState } from '../reducers/user/userReducer';

const initialState = {
  theme: themeIntitialState,
  user: userInitialState,
};

export const loadState = () => {
  const state = localStorage.getItem('job_portal_state');
  if (state === null) {
    return undefined;
  }
  return JSON.parse(state);
};

export const saveState = (state) => {
  const { user, theme } = state;
  const currentState = JSON.stringify({ theme, user });
  localStorage.setItem('job_portal_state', currentState);
};

export const savedState = loadState();

export const loadInititalAppState = () => {
  saveState(initialState)
}