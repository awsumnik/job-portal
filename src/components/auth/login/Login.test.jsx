import React from 'react';
import { Provider } from 'react-redux';
import {BrowserRouter as Router} from 'react-router-dom';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import Login from './Login';
import store from '../../../store/configure-store';

describe('Login Page', () => {
  
  test('should render login page', async () => {
    const { container } = render(
      <Provider store={store}>
        <Router>
          <Login />
        </Router>
      </Provider>
    );
    expect(container).toBeDefined();
    expect(container).toMatchSnapshot();
    const linkElement = screen.getByText(/Log in to Job Portal/i);
    expect(linkElement).toBeInTheDocument();
    const loginButton = screen.getByRole('button', {
      name: /log in/i
    });
    expect(loginButton).toBeDisabled();
  });

  test('should display error for invalid input', async () => {
    render(
      <Provider store={store}>
        <Router>
          <Login />
        </Router>
      </Provider>
    );
    const emailInput = screen.getByPlaceholderText('Enter Email');
    fireEvent.change(emailInput, { target: { value: '12345' } });
    fireEvent.blur(emailInput);
    expect(screen.getByText(/Email should be minimum 8 characters in length./i)).toBeInTheDocument();
  });

  test('login button to be enabled after email and password are inputed', async () => {
    render(
      <Provider store={store}>
        <Router>
          <Login />
        </Router>
      </Provider>
    );
    const emailInput = screen.getByPlaceholderText('Enter Email');
    const passwordInput = screen.getByPlaceholderText('Enter Password');
    fireEvent.change(emailInput, { target: { value: 'anil@test.com' } });
    expect(emailInput.value).toBe('anil@test.com');
    fireEvent.blur(emailInput);
    fireEvent.change(passwordInput, { target: { value: 'Test-123' } });
    expect(emailInput.value).toBe('anil@test.com');
    fireEvent.blur(passwordInput);
    
    const loginButton = screen.getByRole('button', {
      name: /log in/i
    });
    await waitFor(() => {
      expect(loginButton).toBeEnabled();
    });
    fireEvent.click(loginButton);
  });
});