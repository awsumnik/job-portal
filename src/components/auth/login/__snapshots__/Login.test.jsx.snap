// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page should render login page 1`] = `
<div>
  <div
    class="sc-dkzDqf iGZwPq"
  >
    <form
      autocomplete="off"
      class="form"
    >
      <div
        class="sc-gsnTZi eyomTB card"
        style="width: 400px;"
      >
        <div
          class="card-title text-center"
        >
          Log in to Job Portal
        </div>
        <div
          class="card-content"
        >
          <div>
            <div
              class="form-group"
            >
              <label
                for="email"
              >
                Email
              </label>
              <input
                aria-label="Email"
                aria-required="true"
                class="form-input"
                id="email"
                name="email"
                placeholder="Enter Email"
                required=""
                type="email"
                value=""
              />
            </div>
          </div>
          <div>
            <div
              class="form-group"
            >
              <label
                for="password"
              >
                Password
              </label>
              <input
                aria-label="Password"
                aria-required="true"
                class="form-input"
                id="password"
                name="password"
                placeholder="Enter Password"
                required=""
                type="password"
                value=""
              />
            </div>
          </div>
          <div>
            <div
              style="display: flex; justify-content: center;"
            >
              <button
                class="sc-bczRLJ cVLaUc form-button"
                disabled=""
                id="loginBtn"
                name="loginBtn"
                type="submit"
              >
                Log in
              </button>
            </div>
          </div>
          <div>
            <hr />
            <p
              class="text-center"
            >
              Don't have an account?
            </p>
            <div
              style="display: flex; justify-content: center;"
            >
              <button
                class="sc-bczRLJ cBkcFt form-button"
                id="signUpBtn"
                name="signUpBtn"
                type="button"
              >
                Sign Up
              </button>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;
